TYPO3.Flow/Classes/TYPO3/Flow/Package/PackageManager.php

```diff
@@ -741,419 +693,451 @@ public function registerPackage(PackageInterface $package, $sortAndSave = true)
      *
-     * @return void
+     * @param string $packageKey The package key to validate
+     * @return boolean If the package key is valid, returns TRUE otherwise FALSE
+     * @api
      */
-    protected function sortAvailablePackagesByDependencies()
+    public function isPackageKeyValid($packageKey)
     {
-        $sortedPackages = array();
-        $unsortedPackages = array_fill_keys(array_keys($this->packages), 0);
+        return preg_match(PackageInterface::PATTERN_MATCH_PACKAGEKEY, $packageKey) === 1;
+    }
 
-        while (!empty($unsortedPackages)) {
-            reset($unsortedPackages);
-            $this->sortPackagesByDependencies(key($unsortedPackages), $sortedPackages, $unsortedPackages);
+    /**
+     * Resolves package key from Composer manifest
+     *
+     * If it is a Flow package the name of the containing directory will be used.
+     *
+     * Else if the composer name of the package matches the first part of the lowercased namespace of the package, the mixed
+     * case version of the composer name / namespace will be used, with backslashes replaced by dots.
+     *
+     * Else the composer name will be used with the slash replaced by a dot
+     *
+     * @param array $manifest
+     * @param string $packagePath
+     * @return string
+     */
+    protected function getPackageKeyFromManifest(array $manifest, $packagePath)
+    {
+        if (isset($manifest['extra']['neos']['package-key']) && $this->isPackageKeyValid($manifest['extra']['neos']['package-key'])) {
+            return $manifest['extra']['neos']['package-key'];
         }
 
-        $this->packages = $sortedPackages;
+        $composerName = $manifest['name'];
+        $autoloadNamespace = null;
+        $type = null;
+        if (isset($manifest['autoload']['psr-0']) && is_array($manifest['autoload']['psr-0'])) {
+            $namespaces = array_keys($manifest['autoload']['psr-0']);
+            $autoloadNamespace = reset($namespaces);
+        }
 
-        $packageStatesConfiguration = array();
-        foreach ($sortedPackages as $packageKey => $package) {
-            $packageStatesConfiguration[$packageKey] = $this->packageStatesConfiguration['packages'][$packageKey];
+        if (isset($manifest['type'])) {
+            $type = $manifest['type'];
         }
-        $this->packageStatesConfiguration['packages'] = $packageStatesConfiguration;
+
+        return $this->derivePackageKey($composerName, $type, $packagePath, $autoloadNamespace);
     }
 
     /**
-     * Recursively sort dependencies of a package. This is a depth-first approach that recursively
-     * adds all dependent packages to the sorted list before adding the given package. Visited
-     * packages are flagged to break up cyclic dependencies.
+     * Derive a flow package key from the given information.
+     * The order of importance is:
+     *
+     * - package install path
+     * - first found autoload namespace
+     * - composer name
```

Comment:
Just a nitpick, but would be cool to have order of arguments match the order of importance, wdyt?